feat($compile): add $onDestroy directive lifecycle hook

bobbijvoet opened the pull request:
This PR introduces a new $onDestroy lifecycle hook for directives. This hook matches the destroy hook that's present in Angular 2. 

Closes #14020

Commit:
feat($compile): add $onDestroy directive lifecycle hook

Introduce a destroy hook for directives which follows the ng2 component lifecycle

Closes #14020

dcherman commented on src/ng/compile.js:
```diff
@@ -2424,6 +2426,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             controller.instance.$onInit();
           }
+          if (isFunction(controller.instance.$onDestroy)) {
+            scope.$on('$destroy', controller.instance.$onDestroy);
```
Two main comments here (was actually implementing this just now until I saw this):

1.  This doesn't allow late-bound `$onDestroy` methods to be added.  If it's not there at the time of controller instantiation, the lifecycle hook will not function.  Does that follow expectations, or do we need to be checking this at the time of destruction?

2.  The `this` binding in the `$onDestroy` method is going to be incorrect here due to the manner in which you're passing `$onDestroy`.  I see that you copied the unit tests from the one above it as I had done, but removed the `check` call.  You should re-introduce that portion of the test since it would have failed on this point

bobbijvoet replied:
Thanks for your input. I fixed you second point.

Commit:
feat($compile): add $onDestroy directive lifecycle hook

Bind $onDestroy handler to the controller instance

Closes #14020

bobbijvoet commented:
@dcherman Yes, we should allow late-bound $onDestroy methods. We could attach scope.$on('$destroy') listeners for every controller and then check if the $onDestroy method is there. But it does not feel right to attach this listener on every controller. What do you think?

gkalpak commented on src/ng/compile.js:
```diff
@@ -2424,6 +2426,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             controller.instance.$onInit();
           }
+          if (isFunction(controller.instance.$onDestroy)) {
+            scope.$on('$destroy', controller.instance.$onDestroy.bind(controller.instance));
```
The problem with this is that `scope` might be a parent scope that is not directly related to the element.
Maybe we could use `$element.on('$destroy', ...)` instead (I haven't given it much thought, though).

bobbijvoet replied:
Ok. The element.$on solution sounds better. Will create a new solution.

gkalpak commented on src/ng/compile.js:
```diff
@@ -296,6 +296,8 @@
  *   had their bindings initialized (and before the pre &amp; post linking functions for the directives on
  *   this element). This is a good place to put initialization code for your controller.
+ * * `$onDestroy` - Called when the controller's scope has been destroyed. This is a good place to put
+ * code for tearing down your controller and for example removing event listeners.
```
We should also be explicit about when this is called wrt to child component/directives.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -5351,6 +5357,32 @@ describe('$compile', function() {
       });
     });
 
+    it('should call `controller.$onDestroy`, if provided when the controllers is destroyed', function() {
```
_controllers_  -->  _controller_

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -5741,6 +5773,8 @@ describe('$compile', function() {
         siblingController = this.friend;
       };
       spyOn(MeController.prototype, '$onInit').andCallThrough();
+      MeController.prototype.$onDestroy = function() {};
```
Place this line right under `MeController.prototype.$onInit = ...`.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -5788,6 +5826,9 @@ describe('$compile', function() {
       };
       spyOn(MeController.prototype, '$onInit').andCallThrough();
 
+      MeController.prototype.$onDestroy = function() {};
```
Place this line right under `MeController.prototype.$onInit = ...`.

Commit:
feat($compile): add $onDestroy directive lifecycle hook

Call $onDestroy when element is removed, possibility to set $onDestroy method after controller initialization

Closes #14020

dcherman commented on src/ng/compile.js:
```diff
@@ -2419,11 +2421,18 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
             controller.instance.$onInit();
           }
+
+          $element.on('$destroy', function() {
```
The `$destroy` event won't function properly when you have a `transclude: "element"` directive since events are not fired on the comment node that becomes `$element`.

http://jsfiddle.net/rogqp47y/17/

drpicox replied:
So it should be implemented using $scope.$on('$destroy', ... ?

dcherman commented:
Figuring out exactly where to run this logic is trickier than it looks.  We only have two existing hooks that can be used for something like this which are the `$destroy` events fired on both the scope and element.  Both events have disadvantages that make them non-functional in some scenarios which puts us in a tight spot here.  I haven't thought of an alternative solution yet.

dcherman commented:
So I was thinking about this for a little bit this morning, and although the `$destroy` event on the element is close to what we want, it simply won't work for comment nodes.  There's a variety of methods I can think of to kind of emulate that, but none of them are synchronous and match the behavior of the `$destroy` event, and every idea probably has some horrible performance implications as well.

I have a potentially simpler idea though.  Right now, the scope hierarchy is almost entirely managed by higher level scopes;  there's no implicit dependency on the `$destroy` element event being invoked to cleanup the hierarchy.  As a result, that means that anyone that wishes to do anything that results in scope cleanup must already be calling `$destroy` on themselves (including internals like `ngRepeat` and `ngIf`.

With that in mind, what if we were to detect that no scope was requested for the current set of elements, but controllers exist that may potentially contain `$onDestroy` methods?

```js
if ( !directive.scope) {
  var teardownScope = parentScope.$new();

  teardownScope.$on('$destroy', function() {
      // Iterate Controllers here
  });
}
```

This shouldn't affect the user's scope hierarchy since it's a sibling to any existing scopes, similar to a transcluded scope.  You could even expand on this solution and only do this for directives that have a comment node as the root element, although I'd be concerned about potential minor differences between the element lifecycle and the scope one, mainly in the ordering of when the `$onDestroy` methods would be invoked.

I have no data to back up this assertion (at all), but I feel like it may not be that common to have a controller on a directive that doesn't define either an isolate or new scope, so this codepath may be hit less frequently than we might imagine.  The new `component` helper event reinforces that idea by requiring an isolate scope be used.

gkalpak commented:
The more I think about it, the more it doesn't make sense to try to cover all corner-cases and end up with a complicated bloat (that will have little resemblance to ng2).

Since the main purpose (or at least on of the main purposes) of the directive lifecycle hooks is to enable structuring your directive in a way that is close to the ng2-way (and thus facilitate the migration), it is reasonable to only offer the `$onDestroy` hook on directives that define their own scope.

In ng2 all directives have sort of their own scope/context (the component/directive class). So, by not providing the `$onDestroy()` hook on directives with `scope: false`, we won't be hurting the migration (since those directives are not "migratable" anyway).

This way, we will avoid lots of irrelevant corner cases and we'll be able to implement it using `$scope.$on('$destroy', ...)`, which is straight-forward and safe.

WDYT ?

/cc @petebacondarwin

drpicox commented:
I start to think that may be components are not directives:
_(in the same way that squares does not inherit rectangles, ¿how do you change height but not width and still being a square?)_

- they have always isolated scope,
- they are always restrict to entity,
- always have controller (with bindings),
- templates does not have namespace,
- no multielement, no link, ...
- there cannot be two components in the same element (E with controller)

So, may be hooks can only be applied from component constructor or only to components. In such case no need for forEach controller, less overhead in directives, ... (may be all of this can be achieved by providing pre/post-link functions just for components).

petebacondarwin commented:
I think we should implement it via `$scope.$on('$destroy', ...)` for all directives that have a controller which has a `$onDestroy` method. If they are not creating their own scope then they will just have to wait for their containing scope to be destroyed, which should really map exactly to when their element is destroyed.

drpicox commented:
What I mean about components "_are not_" directives is may be that something such simple as changing  `compile.js` to add pre-post link like that:

```javascript
      return {
        controller: controller,
        controllerAs: identifierForController(options.controller) || options.controllerAs || '$ctrl',
        template: makeInjectable(template),
        templateUrl: makeInjectable(options.templateUrl),
        transclude: options.transclude,
        scope: {},
        bindToController: options.bindings || {},
        restrict: 'E',
        require: [name].concat(options.require),
        link: {
          pre: function(scope, element, attrs, controllers) {
            if (isFunction(controllers[0].$onInit)) {
             controllers[0].$onInit();
            }
          },
          post: function(scope, element, attrs, controllers) {
            scope.$on('$destroy', function() {
              if (isFunction(controllers[0].$onDestroy)) {
               controllers[0].$onDestroy();
              }
            });
            // placed after registering to avoid not firing destroy if afterViewInit throws an exception (and save a non-optimizable try-catch)
            if (isFunction(controllers[0].$afterViewInit)) {
             controllers[0].$afterViewInit();
            }
          },   
      };
```

petebacondarwin commented:
@drpicox - I like what you are saying but I feel that developers ought to expect the same hooks to be called even if they create component directives using the `mod.directive()` helper - as long as their directive looks like a component.

drpicox commented:
Yes, you are right. I was just trying to give another vision, but right, it is not the best one.
